File: src/database_interface.ts
/**
 * Database interface with connection pooling for Cloud Run
 * Supports PostgreSQL (production) with proper Cloud Run lifecycle management
 */

import { Pool } from 'pg';
import { getDatabaseUrl, getUserManagementDatabaseUrl } from './config';
import logger from './logger';

export class DatabaseInterface {
    private pool: Pool;
    private databaseUrl: string;
    private ended = false;

    constructor(databaseUrl: string) {
        this.databaseUrl = databaseUrl;
        this.pool = new Pool({
            connectionString: databaseUrl,
            ssl: {
                rejectUnauthorized: false,
            },
            // Connection pool settings optimized for Cloud Run
            min: 1,  // Minimum connections
            max: 3,  // Maximum connections (Cloud Run friendly)
            idleTimeoutMillis: 300000,  // 5 minutes
            connectionTimeoutMillis: 5000,
            // Keep-alive settings
            keepAlive: true,
            keepAliveInitialDelayMillis: 0,
        });
    }

    async connect(): Promise<void> {
        try {
            // Test the connection
            const client = await this.pool.connect();
            await client.query('SELECT 1');
            client.release();
            logger.info('Connected to PostgreSQL database and created connection pool');
        } catch (error) {
            logger.info(error, 'Failed to connect to the database');
            throw error;
        }
    }

    async disconnect(): Promise<void> {
        if (this.ended) {
            return;
        }
        this.ended = true;
        await this.pool.end();
    }

    async reconnect(): Promise<void> {
        logger.info('Forcing database reconnection...');
        await this.disconnect();
        // Re-initialize the pool
        this.pool = new Pool({
            connectionString: this.databaseUrl,
            ssl: {
                rejectUnauthorized: false,
            },
            min: 1,
            max: 3,
            idleTimeoutMillis: 300000,
            connectionTimeoutMillis: 5000,
            keepAlive: true,
            keepAliveInitialDelayMillis: 0,
        });
        this.ended = false;
        await this.connect();
    }

    async execute(query: string, params: any[] = []): Promise<any> {
        const client = await this.pool.connect();
        try {
            if (query.trim().toUpperCase().startsWith('SELECT')) {
                const result = await client.query(query, params);
                // Return results and column names for consistency
                const columnNames = result.fields.map(field => field.name);
                const rows = result.rows.map(row => Object.values(row));
                return { rows, columnNames };
            } else {
                // For non-SELECT queries, return the number of affected rows
                const result = await client.query(query, params);
                return result.rowCount || 0;
            }
        } catch (error) {
            logger.info(error, 'Error executing query', { query, params });
            throw error;
        } finally {
            client.release();
        }
    }

    async executeOne(query: string, params: any[] = []): Promise<any> {
        const client = await this.pool.connect();
        try {
            const result = await client.query(query, params);
            return result.rows[0] || null;
        } catch (error) {
            logger.info(error, 'Error executing query', { query, params });
            throw error;
        } finally {
            client.release();
        }
    }

    async isHealthy(): Promise<boolean> {
        if (!this.pool) {
            return false;
        }
        try {
            const result = await this.pool.query('SELECT 1');
            return result !== null;
        } catch (error) {
            logger.error(error, `Health check failed for ${this.databaseUrl}`);
            return false;
        }
    }
}

// Global database instances
export const monitoringDb = new DatabaseInterface(getDatabaseUrl());
export const userManagementDb = new DatabaseInterface(getUserManagementDatabaseUrl());

File: src/database_interface.py
"""
Simple database interface using raw SQL
Supports PostgreSQL (production)
"""

import logging
from typing import Optional, Dict, Any
from contextlib import asynccontextmanager

try:
    import asyncpg
    POSTGRES_AVAILABLE = True
except ImportError:
    POSTGRES_AVAILABLE = False

from src.config import get_database_url, get_user_management_database_url

logger = logging.getLogger(__name__)


class DatabaseInterface:
    """Simple database interface using raw SQL"""

    def __init__(self, database_url: str = None):
        self.database_url = database_url or get_database_url()
        self.connection = None

    async def connect(self):
        """Connect to the database"""
        if not POSTGRES_AVAILABLE:
            raise ImportError("asyncpg is required for PostgreSQL support")
        
        self.connection = await asyncpg.connect(self.database_url)
        logger.info(f"Connected to PostgreSQL database")

    async def disconnect(self):
        """Disconnect from the database"""
        if self.connection:
            await self.connection.close()
            self.connection = None

    @asynccontextmanager
    async def get_connection(self):
        """Get a database connection (context manager)"""
        if not self.connection:
            await self.connect()
        yield self.connection

    async def execute(self, query: str, params: tuple = ()) -> Any:
        """Execute a query and return results"""
        # Normalize query for the current database type
        normalized_query = self.normalize_query(query)

        async with self.get_connection() as conn:
            if query.strip().upper().startswith("SELECT"):
                results = await conn.fetch(normalized_query, *params)
                # For consistency with SQLite, return results and column names
                if results:
                    # Get column names from the first row
                    column_names = list(results[0].keys()) if results else []
                    # Convert asyncpg.Record objects to tuples for consistency
                    tuple_results = [tuple(row.values()) for row in results]
                    return tuple_results, column_names
                else:
                    return [], []
            else:
                # For non-SELECT queries (INSERT, UPDATE, DELETE), return the command status
                # which indicates the number of affected rows
                result = await conn.execute(normalized_query, *params)
                # asyncpg returns a string like "UPDATE 1" or "INSERT 0 1"
                # We need to extract the number of affected rows
                if isinstance(result, str):
                    # Extract the number from strings like "UPDATE 1" or "INSERT 0 1"
                    import re
                    match = re.search(r"(\d+)$", result)
                    return int(match.group(1)) if match else 0
                return result

    async def execute_one(self, query: str, params: tuple = ()) -> Optional[Dict]:
        """Execute a query and return one result"""
        # Normalize query for the current database type
        normalized_query = self.normalize_query(query)

        async with self.get_connection() as conn:
            row = await conn.fetchrow(normalized_query, *params)
            return dict(row) if row else None

    def normalize_query(self, query: str) -> str:
        """Normalize query for PostgreSQL"""
        # PostgreSQL uses $1, $2, etc. for parameters
        # This is already the format we expect, so just return as-is
        return query


# Global database instances
monitoring_db = DatabaseInterface(get_database_url())
user_management_db = DatabaseInterface(get_user_management_database_url())
